numpy/core/src/_simd/_simd_inc_arg.h

```diff
@@ -0,0 +1,68 @@
+#ifndef NPY_SIMD
+    #error "Not a standalone header"
+#endif
+
+/************************************
+ ** Protected Definitions
+ ************************************/
+static int
+simd_arg_parse(PyObject *py_args, simd_arg *args, int args_n)
+{
+    Py_ssize_t cur_arg = 0;
+    if (py_args == NULL) {
+        if (args_n == 0) {
+            return 1;
+        }
+        goto wrong_size;
+    }
+    int arg_pos = 0;
+    if (!PyTuple_Check(py_args)) {
+        if (args_n != 0) {
+            goto wrong_size;
+        }
+        simd_arg *a = &args[0]; a->obj = py_args;
+        a->data = simd_obj2data(py_args, a->type);
+        if (PyErr_Occurred()) {
+            goto wrong_data;
+        }
+        return 1;
+    }
```

Comment:
Can this ever happen? Python will never pass a non-tuple into `args` for a C extension function, and if you're calling one yourself that way, don't.

Reply:
> Can this ever happen? Python will never pass a non-tuple into args for a C extension function

Yes, when it passes only "one" argument 

>  and if you're calling one yourself that way, don't.

sorry, I didn't get you. Do you mean (my just in case rule :) )?

Reply:
> Yes, when it passes only "one" argument

Can you point me to the line where this happens?

> and if you're calling one yourself that way, don't.

I mean if you are manually calling a function with signature `PyObject *module, PyObject *args`. Those are typically only called by python, not by other C code.

Reply:
> Can you point me to the line where this happens?

`simd_arg_parse` is directly parsing the PyObject* from the module method's args, without adding PyArg_ParseTuple or any of its family in between, which makes it too fast and easy to debug(gdb).

Again Python passes a single PyObject* if the call carries only one argument and a tuple of PyObject* if the call
holds multiple arguments. I'm not sure what their documents say but their code say that.

Reply:
> Again Python passes a single PyObject* if the call carries only one argument and a tuple of PyObject* if the call
holds multiple arguments.

If that were true, python would not be able to distinguish `f((1,2))` and `f(1,2)`, which is obviously untrue

Reply:
> I mean if you are manually calling a function with signature PyObject *module, PyObject *args

No, I assign the pointer of C function directly to `.m_methods`

Reply:
> Python passes a single PyObject* if the call carries only one argument

I think you're thinking of the behavior of `METH_OLDARGS`, which is now unsupported, and you're not using anyway.

Reply:
> I think you're thinking of the behavior of METH_OLDARGS, which is now unsupported, and you're not using anyway.

My bad, It seems I get confused from their backward support.